### PR TITLE
Added the capability to archive to Azure Blob Storage

### PR DESCRIPTION
**Pull request by DayoOliyide**
Hi,

This pull request is to add Azure Blob storage capabilites to Bifrost.
We had a need to backup all kafka messages to the Azure Blob Storage, and we made some changes
to Bifrost to make this possible.
Over time we found this to be very useful, and it's allowed us to easily recover from a few issues.
Thus, we would like to share this extra capability, as we suspect there are others who might
find this useful.

Thanks
Dayo & Bruno

### CODE REVIEW

**Comment by pingles** [comment]
Wow, thank you! I'll try and get this merged today/tomorrow.

**Comment by BrunoBonacci** [comment]
Hi @pingles ,
I worked with Dayo in the same project and the extension has been working in production for over 6 months with no issues.
We decided to make a few changes to the name of the files which were uploaded to the bucket for a number of reasons. The initial file name contained only the first offset. One problem that we faced was to understand if all "chunks" (baldr files) were uploaded in the deep storage (S3 or Azure blob) and make sure that there were no gaps such as file uploads which failed for some reason and we failed to notice it.
The only way to verify this was to read all files, finding out last message offset and checking that the next file was starting with the next offset. This was very impractical and time consuming.
We decided to add the last offset as part of the name which gave us the possibility to run a script, and just by looking a the name of the files, determine if there was any file missing. In a corporate environment assuring that your backups are complete and valid is a major point.
Finally we decided to add the timestamp as part of the file (as milliseconds from epoc) due an error we experienced during the Azure Global outage of last November (http://azure.microsoft.com/blog/2014/11/19/update-on-azure-storage-service-interruption/).
Due to this outage we had two out of 3 brokers killed during the outage and the third one having bad time. During the leader rebalancing and the brokers dying, we had a OffsetOutOfRangeException (if I well recall) and bifrost, for a few partitions, reset the offset counter to the "earliest" offset available in the log as per our configuration. By doing so bifrost started to re-consume all messages from the beginning of the log and push them to the deep storage. At that point the file names were containing only "start-stop" offsets and we observed overlapping offsets ranges which made really hard to reprocess the data avoiding a large amount of duplicates. Due to this issue, we then decided to add the uploading timestamp as first parameter **"TS${timestamp}-${start-offset}-${stop-offset}"** which gives the following advantages:
  - possibility to see gaps (missing messages) automatically by just looking a the file names
  - possibility to avoid duplicates when offset-reset occurs in consumers just by looking at the file names

If you want more detail/explanation about these changes I'm happy to set up a meeting the three of us.

Bruno

**Comment by tgk** [comment]
Impressive story! It's great to see others have had some experience with the platform system.

I think the naming convention is an interesting point. Naming the files as `first-offset.baldr.gz` was done to allow any step of the upload process to fail and be retried as an idempotent operation. The classic example is:

- File `i.baldr.gz` is uploaded with messages `i, ..., i+n`
- Catasrophic error occurs before the offset `i+n` is committed to Zookeeper
- Bifrost restarts, reprocesses messages from offset `i`.
- File `i.baldr.gz` is uploaded with messages `i, ..., i+m`, 'n != m', overwriting existing file.
- `i+m` is committed to Zookeeper

In the above scenario, if the new naming scheme had been used the new file would not have overwritten the existing file, and the same messages would be present in multiple files in S3/Azure blob storage. The change in naming convention invalidates the idempotency guarantees.

We did not anticipate the error you've described and clearly this naming strategy breaks down in that case. I would still be hesitant in adopting it as the default though, as it breaks the idempotency property of Bifrost.

We had a presentation at last years Clojure eXchange describing some of the infrastructure points of Bifrost. If memory serves there is a brief discussion around the file naming convention and the benefits, but probably nothing I haven't mentioned here. The video can be found [here](https://skillsmatter.com/skillscasts/6032-more-open-source-systems-please).

Once again, thank you so much for your pull request, and thanks for the excellent explanation. I'd like to know what @pingles thoughts are :-)

**Comment by pingles** [comment]
Thanks for your work and sharing the story- I definitely think adding log verification would be a useful thing to support but, as Thomas says, I think I prefer the current naming model for idempotent writes. But, I probably need to think more about it.

IIRC we only persist the message payload in the baldr archive so there's no offset data stored within the archives themselves so you could only verify from the file names.

Definitely need to think about this a bit more- I'd really like to keep the simplicity and idempotence of what we do now but additionally support verification.

Having said that, adding Azure storage is awesome- thanks for going through separate commits :)

**Comment by BrunoBonacci** [comment]
Hi @tgk,
In the example you provided even the current naming convention wouldn't assure idempotency in all cases. In fact in the case we described you will end up with duplicate overlapping baldr files with near impossible way to distinguish which one are original and which ones are duplicates (the only way is to use the upload metadata provided by the platform S3/Azure).

So the question is: _if you are not able to provide idempotency in face of any type of failures, would you still consider this a property of your system (upload process part)?_

Since I know that this cannot be guaranteed, I prefer to consider the deep storage (S3/Azure blob/HDFS) in this context as a immutable storage. So once you upload something, this should never be changed (nor overwritten with different content).
The proposed naming convention offer the possibility to defise a natural sequence of files and easily understand if segments are duplicated or missing. 
With the current naming convention, how would you answer the question: "*Are you sure that all messages are backed up into deep storage?*". To put in place a process able to answer this question is quite a laborious process. With the naming convention we proposed it is very easy to write a script which verifies that. You could even remove duplicate segments generated by your failure example automatically.
This could be even a nice addition to do to bifrost. `:verify-upload-contiguity true` and `:auto-clean-duplicate-segments :true` where a background thread would look at the sequentiality of the offsets for missing chunks (and upload them when a gap is found) and duplicates overlapping segments and remove them.

**Comment by tgk** [comment]
Hi @BrunoBonacci 

I think you make a very valid and strong point, namely that any type of failure might arise. We built a very small and simple product (Bifrost) with what we feel are very good guarantees because we rely on the assumption that Zookeeper will be consistent (and that S3 doesn't spontaneously remove objects). That is the only assumption that we would like to rely on. You are suggesting that we should abandon that assumption (which I totally understand) but I wonder if other parts of Bifrost would need to be reevaluated then. The simplicity of Bifrost is there only because of one hard assumption: Zookeeper is sound. You have given empiric evidence that that is not always a great assumption.

Sorry about the massive wall-of-text. I really, really appreciate your feedback and thoughts. I think my thinking around Bifrost is very biased towards trusting Zookeeper and the offsets stored there, mainly because I want to centralise concerns around consistency. Again, you have given me (and @pingles) food for thought.

**Comment by BrunoBonacci** [comment]
My Kudos to you guys for making bifrost, @DayoOliyide can confirm, it has been our safety net a few times.

Regarding Zookeeper consistency, I don't want to imply that it has any problem. If you wish I can describe in better details the failure mode and why (i think) this happened.

**Comment by pingles** [comment]
Thanks, very glad to hear it's been useful for you. It would be good to
hear more about what happened if you could share please.

**Comment by BrunoBonacci** [comment]
I believe we are all in London, so, if you prefer, we could meet somewhere share the details, maybe in front of a pint. Otherwise I will need a bit of time to write out all necessary details.

**Comment by tgk** [comment]
@BrunoBonacci @pingles I vote for pint based communication.

**Comment by pingles** [comment]
Sounds great. Perhaps somewhere around St Pauls next week/week after?

**Comment by BrunoBonacci** [comment]
St.Paul works well for me too. What about Wednesday the 22nd after office time (18.00 circa)?

**Comment by tgk** [comment]
I'm hosting a golang meetup on the 22nd so I'll have to opt out. What about Monday?

**Comment by BrunoBonacci** [comment]
Monday the 20th same time it's fine.

**Comment by tgk** [comment]
I spoke too soon - it appears I have other plans on Monday. Sorry.

Basically, I wouldn't rely too much on me. If you guys are happy to meet up on Wednesday go ahead without me. Otherwise next Friday might be good.

Sorry about the confusion.

**Comment by pingles** [comment]
Haha, Friday should be fine for me- Bruno?

**Comment by BrunoBonacci** [comment]
Any day is fine for me next week a part from Tuesday. So I guess let's call it for Friday 24th? @DayoOliyide what about you?

**Comment by DayoOliyide** [comment]
Friday is fine for me.

To echo Bruno, we were only able to recover (data wise) from the above service interruption due to bifrost. So Kudos

I'll dig around to see if I can find any more information about what we noticed about the unique circumstances we found ourselves in as a result of the service interruption.

**Comment by BrunoBonacci** [comment]
Hey guys,
any suggestion for the meeting place/pub for tonight?

**Comment by tgk** [comment]
I haven't seen @pingles around today and my child seems to have come down with a virus. I know it's really annoying, but how would you guys feel about postponing?

**Comment by BrunoBonacci** [comment]
no problem for me

**Comment by tgk** [comment]
Hereby postponed then! Any suggestions?

**Comment by tgk** [comment]
Hi @BrunoBonacci and @DayoOliyide 

Once again, it was great seeing you guys at the Experience Report meetup.

Sorry for taking such a long time to get around to this, but I've finally had time to think a bit more about this. There's two changes I'd like before accepting the pull request - I wonder if you'll accept them :-)

1. Do not rename the S3 keys. We'd like to retain the idempotency (at the danger of the error scenario you've outlined (please see my comment below for an alternative suggestion)).
2. Bump the version in VERSION to 0.1.1. The change in the configuration file is, as far as I can tell, a breaking API change to it'd be nice to reflect this in the version number.

I don't know if @pingles has any other comments.

Would you guys be happy with making those changes? We really appreciate the Azure support and we'd like to get it in there.

On the naming convention and the potential problems in a catastrophic failure like the one you've experienced I propose adding support for a new upload strategy:

Rather that using sequence files (such as `baldr`) individual messages are uploaded to S3 or Azure using the filename convention:

    <id>-<sha of content>.raw

This means that stores are still idempotent but that catastrophic failures where `id`s are reset to an earlier version does not overwrite data that has already been backed up. The cost is more frequent uploads and that messages can not be downloaded in bulk. An advantages is that it's easier to detect gaps and do random reads of old messages.

I can't figure out if that's a good thing to add to Bifrost or if this is really a completely different application with much simpler semantics. Any thoughts on this point would be highly appreciated.

**Comment by BrunoBonacci** [comment]
Hi @tgk,
are you suggesting to store individual messages?

If so I do agree that this would solve some of the issues highlighted earlier, however it introduces an unnecessary burden to the deep storage platform.
The current solution already creates too many small files. Platforms such as [Amazon S3](http://blog.mortardata.com/post/58920122308/s3-hadoop-performance), Azure Blob  Storage  and HDFS are known to perform poorly in presence of huge amount of small files.
This is not only true for pushing the data, but especially true for reading and reprocessing the data.
By storing individual messages this problem will become more evident as suddenly one or two orders of magnitude more files will need to be stored.
One of the things I've done in my project to avoid this issue is to do a Cascalog job which unpack and regroup messages in big files (1 file x hour x partition). Larger files are way more efficient on DFS systems.

I had some thought on the discussion we had just before leaving the Clojure event.
Maybe  having Bifrost to store fixed amount of messages per file is the best of both world, but it will need an addition.
For example we could say that we want to pack messages in bulk of 10K., bifrost could keep accumulating until this threshold is reached, and the push the message as:
`/<topic>/<partition>/<offset_start>-<offset-stop>_[maybe_sha].baldr.gz`
Now this will satisfy the idempotency and the possibility to find gaps as well. The offset-stop is required in case the buffer size changes, the SHA can be a nice addition.
Now the issue with this solution is that the risk of permanent loss of messages in case of catastrophic failure is higher than the current solution where messages are pushed into a higher durability store in seconds after the arrival.
To mitigate this problem a possibility could be to save individual messages (or small batches) in the deep storage in parallel to the local buffering. Something like:
`/<topic>/<partition>/RAW_<offset_start>-<offset-stop>/<id>.gz`

Now when the buffer fills up and it is successfully pushed to deep storage, a command can be issued to delete the RAW events. This solution would be able to cover the case in which the buffer is not yet full (and only stored in the bifrost machine) and a major failure occurs.

**Comment by tgk** [comment]
@BrunoBonacci Interesting - I completely agree, S3 and the like should really contain big files for processing purposes. I like your separation between deep storage and raw files. It feels like two different steps now.

This is becoming a long thread, but it's very interesting. A more complicated system would be something like:

    Kafka <- [Application 1] -> [shallow storage] <- [Application 2] -> [deep storage]

That is, there is no reason why one application should handle both shallow and deep storage. Your shallow trusted storage could be Riak/Dynamo or another AP system that is more trusted than Zookeeper + Kafka.

Mo moving parts, mo problems though.

Again, thanks for all your thoughts! Just as an update, one of our teams is looking into detecting holes in and corruption of the data that makes it way to S3 and how to detect that which is somewhat related to our discussion :-)

**Comment by BrunoBonacci** [comment]
I agree, adding another system would just make the solution more complex, I'd rather keep it simple.

What i was suggesting is to keep the solution as it is and have a "fast" lane messages are pushed every 30s/1m (just like it is now) and add a "slow" lane which packs messages in larger batch (100K..10M)
and when the large batch is pushed to the same deep storage of the "fast" lane then you can get rid of the fast lane messages.

```

                                           |   Kafka
                                           |
                                           |
                                           |
                                           |
                                +----------v-----------+
                                |                      |
                                |  bifrost consumer    |
                                |                      |
                                +----------+-----------+
                                           |      
                                           |      
                                           |      
                     fast lane             |     slow lane
                 +-------------------------+---------------+
                 |                                         |
                 |                                         |
               +-v-+                              +--------v-------+
    local fs   |   |                              |                |  local fs buffer
    buffer     |   |                              |  from       0  |  it buffers 1M msg
    which is   |   |                              |  to   1000000  |  and the push to  
    pushed     +---+                              |                |  deep storage   
    frequen.   +---+                              |                |  when buffer is     
    ~30s/1m    +---+                              |                |  full.
                 |                                |                |       
                 |                                |                |       
                 |                                +----------------+       
                 |                                +--------+-------+
                 |                                         |
     /topic1/0/fast/00000000-01000000/001230.bladr.gz      |
                 |                                         |
                 |                                         |  
                 |                                 /topic1/0/00000000-01000000.bladr.gz
                 |                                         |                           
             +---v-----------------------------------------v--------+                  
             |                                                      |                  
             |                   Deep Storage                       |                  
             |                                                      |                  
             |       Amazon S3        /     Azure Blob storage      |                  
             |                                                      |
             |                                                      |
             +------------------------------------------------------+
```

The current functionality will become the fast lane, a new buffer will be added to implement the slow lane in the exact same way the fast lane works, with the only difference that a new buffer will be created not based on the time but on the configurable number of events.

Once the big buffer containing all the messages for a given partition between `offsetA` and `offsetB` is pushed to the deep storage than a `rm -fr /topic1/0/fast/$offsetA-$offsetB/*` can be issued on the deep storage for the *fast lane* to remove all the small messages.

Advantages:
  * Simple, not much different than current behaviour
  * Message gaps are traceable
  * Big files in deep storage
  * Adding the SHA you get the protection against offsets rewind.
  * Same level of data loss risk as now
  * no new moving parts.
  * Size of the big buffer will be configurable as needed

**Comment by tgk** [comment]
That sounds like a pretty cool idea. Would it make sense to just have two instances of Bifrost running in parallel, one configured with time, one with number of messages? I'm just trying to limit the scope of change here and have as simple an application as possible :-)

**Comment by BrunoBonacci** [comment]
From a coordination perspective it would be easier to run them in a single process.
The only reason the "fast last" has to exists is to narrow down the risk of losing messages while the "slow lane" is buffering large batches. Once the large batch has been pushed to S3 there is no more reason to keep the same data in small batches, therefore you can delete it. Due to the coordination involved between the "slow lane" and the "fast lane" I believe it is easier to keep them in a single process.
In other words once the slow lane pushed successfully the buffer `001000000-005000000.bladr.gz` should communicate to the `fast lane` that it can now discard all messages between the same range.
If we keep the "fast lane" directory names as described earlier this can be done with a single remote delete request.

**Comment by DayoOliyide** [comment]
Hi,
Apologies for my silence.
I think this shouldn't be part of this pull, as it is addressing a change in the behaviour of bifrost and isn't specific to Azure Blob Storage. This should be implemented as a new pull/feature.

So I'll carry out the actions as suggested by Thomas to get this pull accepted/closed

- I'll add a commit to keep the old naming convention across both s3 and azure. i.e <offset>.baldr.gz
- Bump the Version number up to Version 0.1.1

Lastly, I just want to double check that everyone is happy with the new configuration that is included
as part of [7866d18](https://github.com/DayoOliyide/bifrost/commit/7866d188e51a986bec739027ee65de530be40f1e)

**Comment by tgk** [comment]
Thanks @DayoOliyide ! I'm more than happy for the new configuration format.

Yes, I think we're moving into a larger discussion around the semantics of Bifrost. It's quite good fun but we're probably moving away from the topic.

**Comment by DayoOliyide** [comment]
Hi,
Just wanted to chase up on this. 
I think I've carried out all the actions required.

Is there anything else you would like me to change ?
-- Dayo

**Comment by tgk** [comment]
Right you are! Sorry about the massive delay. I shall merge :-)
